FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -65,4 +81,198 @@ object PgpMsg {
   )
 
   private val maxTagNumber = 20
+
+  enum class DecryptionErrorType {
+    KEY_MISMATCH,
+
+    // skip, effective for symmetric encryption
+    // USE_PASSWORD,
+
+    WRONG_PASSWORD,
+    NO_MDC,
+    BAD_MDC,
+
+    // a passphrase was not provided for the key which requires it,
+    // not sure how to detect this in the PGPainless.
+    NEED_PASSPHRASE,
+
+    FORMAT,
+    OTHER
+  }
+
+  data class DecryptionError(
+      val type: DecryptionErrorType,
+      val message: String,
+      val cause: Throwable? = null
+  )
+
+  data class DecryptionResult(
+      // provided if decryption was successful
+      val content: ByteArrayOutputStream?,
+
+      // true if message was encrypted.
+      // Alternatively false (because it could have also been plaintext signed,
+      // or wrapped in PGP armor as plaintext packet without encrypting,
+      val isEncrypted: Boolean,
+
+      // pgp messages may include original filename in them
+      val filename: String?,
+
+      // todo later - signature verification not supported on Android yet
+      val signature: Any?,
+
+      // provided if error happens
+      val error: DecryptionError?
+  ) {
+    companion object {
+      fun withError(
+          type: DecryptionErrorType,
+          message: String,
+          cause: Throwable? = null
+      ): DecryptionResult {
+        return DecryptionResult(
+            content = null,
+            isEncrypted = false,
+            filename = null,
+            signature = null,
+            error = DecryptionError(type, message, cause)
+        )
+      }
+
+      fun withCleartext(cleartext: ByteArrayOutputStream, signature: Any?): DecryptionResult {
+        return DecryptionResult(
+            content = cleartext,
+            isEncrypted = false,
+            filename = null,
+            signature = signature,
+            error = null
+        )
+      }
+
+      fun withDecrypted(content: ByteArrayOutputStream, filename: String?): DecryptionResult {
+        return DecryptionResult(
+            content = content,
+            isEncrypted = false,
+            filename = null,
+            signature = null,
+            error = null
+        )
+      }
+    }
+  }
+
+  @Suppress("ArrayInDataClass")
+  data class SecretKeyInfo(
+      val armored: String? = null,
+      val keyRing: PGPSecretKeyRing = readArmored(armored!!),
+      val passphrase: CharArray? = null
+  ) {
+    companion object {
+      private fun readArmored(armored: String): PGPSecretKeyRing {
+        return JcaPGPSecretKeyRingCollection(
+            ArmoredInputStream(armored.toByteArray().inputStream())
+        ).keyRings.asSequence().first()
+      }
+    }
+  }
+
+  fun decrypt(
```

Comment:
It seems this method uses `ByteArray` as an input. 

And here we use RAM twice.

```
val output = ByteArrayOutputStream()
decryptionStream.use { it.copyTo(output) }
```

@tomholub I think the current approach is enough to drop Node.js (I mean don't care about RAM). But it has performance issues for Android. We will have to improve that in the near future(upcoming releases).

Reply:
That's ok, optimizations can be done in a follow-up. Thanks for mentioning - you can file an issue with some things to look into afterwards.

Reply:
> And here we use RAM twice.

Sorry, not twice. Just more.

Reply:
Let's do it later. Added #1219